packages/consensus/src/validation/difficulty.cairo

```diff
@@ -61,60 +61,43 @@ pub fn adjust_difficulty(
     }
 }
 
-/// Convert target value to the compact form (bits)
-fn target_to_bits(target: u256) -> Result<u32, ByteArray> {
-    if target == 0 {
-        return Result::Err("Target is zero");
-    }
-
-    if target > MAX_TARGET {
-        return Result::Err("Exceeds max value");
-    }
-
-    // Find the most significant byte
-    let mut size: u32 = 32;
-    let mut compact = target;
-
-    // Count leading zero bytes by finding the first non-zero byte
-    while size > 1 && shr(compact, (size - 1) * 8) == 0 {
-        size -= 1;
+/// Reduces target precision leaving just 3 most significant bytes.
+///
+/// Note that the most significant byte might be ZERO in case the following
+/// one is >= 0x80 (see "caution" section on https://learnmeabitcoin.com/technical/block/bits/).
+/// This helper assumes the given target is strictly less than MAX_TARGET.
+fn reduce_target_precision(target: u256) -> u256 {
+    let mut shift: u32 = 232; // 29 * 8 for MAX_TARGET
+    while shift != 0 {
+        // Shifting right trying to find the first non-zero byte
+        let msb = shr(target, shift);
```

Comment:
It involves a lot of shr, so don't we want to avoid binary shifts? Or add a method that does binary shifts 8 by 8?

Reply:
We will address shifts later, we can probably leverage static lookup tables since we know our bounds.  
For now it's a good abstraction.

Reply:
But yeah, it's much faster without shifts